### DISCUSSION TITLE
Excel Integration

### DISCUSSION THREAD

**Discussion by ghost** [fattureincloud/api#309] [Ideas]
I created a macro in Excel that takes input from the Excel sheet and from FIC to create a new electronic invoice and submit it to FIC.

It is very basic but makes my administration a lot easier.

I coded everything in VBA, I did not use a FIC SDK.
An alternative would be to us a COM component by wrapping the .NET SDK 

Having a maintained SDK that can be used from within Microsoft Office products (Excel, Access, ...) might be valuable for more clients.

I would be happy to share my code to get this going or to brainstorm other solutions.

**Comment by valmoz** [comment]
Hi @EmondW, that's excellent news!
As a starter, I suggest you create a GitHub public repository and share the link to your repo with us, it could be valuable for other users. 

As explained in the [OpenAPI specification](https://developers.fattureincloud.it/docs/openapi-specification) page in our repository, we don't actually write our own SDKs, but we generate them using the [OpenAPI Generator](https://openapi-generator.tech/). This lets us keep all the SDKs updated by updating the specification. Of course, we personalize the templates as needed to make them work properly and we test the generated result.

Unfortunately, the OpenAPI doesn't provide a VBA generator, so we have two alternatives:
- We provide another Quickstart, as we already did with the [Excel Power Query Quickstart](https://developers.fattureincloud.it/docs/quickstarts/powerquery-quickstart/), to explain how to get started with FIC & VBA
- We implement a totally new SDK without the generator

This second option is really unlikely, so I think the best solution could be for other users to access or even contribute to your code in a public repo. Also, once we review the code we could eventually share the link to your repo in our documentation in a dedicated quickstart.

Let us know what you think about that.

Thank you again for your effort.
Mauro

**Comment by ghost** [comment]
Thanks @valmoz for your reply.

I see one more option: write a VBA Generator. Unfortunately my java skills are not sufficiently up to par to do that.

[Office Script](https://learn.microsoft.com/en-us/office/dev/scripts/overview/excel) (typescript) might also be an option but this seems only available for more expensive licenses of Microsoft Office 365

I will setup a repo add share it with you.

**Comment by ghost** [comment]
As an alternative, it might be possible to change/add the templates for C# and mark the classes as ComVisible.

This way a user could register the library on their system and call the the C# components from VBA.

A lot depends on what these classes look-like but that way it would be a maintained API client.

**Reply by valmoz** [reply]
Ok @EmondW, thank you really much.
I'll share this idea with my colleague because unfortunately I'm not a C#/VBA expert.

For sure we'll not implement a generator from scratch, it would be too costly. It could be something useful for the wider OpenAPI Generator community maybe.

For now, let's start with the repo. I think it would be enough for smaller projects.

Thanks 
Mauro

**Comment by ghost** [comment]
I found a better way than using VBA.
I am building an Excel Add-In using the C# SDK.

So this will no longer be an integration with VBA.

The Add-In project might still be interesting for others. 
Are you still interested in this?

**Reply by valmoz** [reply]
Hi @EmondW 
of course, I think it could be useful to share it with other users. Of course, the add-in could be more complete because it uses our SDK.

Still, if you already created a test with VBA, it could be useful for users even if just for one method. Some people prefer to use VBA instead, so even a fragment could be interesting.

If you already created some VBA code and don't want to open a new repo for it, you can just share it in the discussion.

Thanks
Mauro

**Comment by ghost** [comment]
This is the current state of the Add-in:
![image](https://user-images.githubusercontent.com/105197319/233860815-1a58caff-673c-4c50-ae6d-ca2e0c02f0bf.png)

- The Tab in the Ribbon opens an Action Panel. The Connect button initiates a connection to the API retrieves a list all companies of the user. It also caches company-independent data (languages, currencies, ...)
- After selecting a company and clicking "Configure Invoice", company related data is cached (clients, payment methods, payment accounts, ...) and the two buttons in the ribbon are also enabled.
- The two buttons in the ribbon allow the user to select a cell and then insert a VAT id + description or a Payment Account Id and Description.
- The Line Items and Payments boxes in the panel allow you to select the data for the Items and Payments
- All combo boxes in the panel and in the Insert dialogs support a preference setting. If the cell with the correct name is set in the sheet, the first item that matches with the text in the cell will be selected in the combo box.
- The Upload Invoice is nearly done. It will create an Issued Document/Invoice. Still to be sent to the SDI.

**Reply by emajo** [reply]
Hi @EmondW thanks a lot for the update, your work will be very useful for the community!

**Reply by valmoz** [reply]
Hi @EmondW 
how is the integration going? Do you have some questions for us?

Thanks 
Mauro

**Reply by ghost** [reply]
I am currently testing. The code works for me but I need to strip out and clean up so it is more obvious for others what it is doing and how to use it.

I will provide a VBA sample and an Add-In sample.

**Reply by valmoz** [reply]
that's amazing! thank you for your effort!
